modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpPipeliningHandler.java

```diff
@@ -227,8 +278,28 @@ private boolean doFlush(ChannelHandlerContext ctx) {
         }
         boolean needsFlush = true;
         while (channel.isWritable()) {
-            final WriteOperation currentWrite = queuedWrites.poll();
+            WriteOperation currentWrite = queuedWrites.poll();
             if (currentWrite == null) {
+                drainQueuedNowWritable(ctx);
+                if (channel.isWritable() == false) {
+                    // try flushing to make channel writable again, loop will only continue if channel becomes writable again
+                    ctx.flush();
+                    needsFlush = false;
+                }
+                if (channel.isWritable() == false) {
+                    break;
+                }
+                currentWrite = queuedWrites.poll();
+            }
+            if (currentWrite == null) {
+                // no write from a non-chunked message was found queued to be written, check if a chunked message might have become
+                // writable
+                if (currentChunkedWrite != null) {
+                    if (writeChunk(ctx, currentChunkedWrite.combiner, currentChunkedWrite.response.body())) {
+                        finishChunkedWrite();
+                    }
+                    continue;
```

Comment:
Do we not need a flush here too if the channel is not writeable?

I wonder if we can simplify the logic here, perhaps splitting into 3 methods with the 3 concerns (outer queue, queued writes and chunked writes)?

Reply:
heh I knew you'd notice :) I think the whole business if flushing in the loop is nonsense for normal messages right now and just a waste of cycles. I didn't add it to the chunked write loop on purpose and was going to remove it for normal writes in a follow-up.
Calling flush here will never make the channel writable again, all it will do is enqueue a redundant flush task on the executor. If we aren't writable any more we should exit the loop, pass a single flush along and wait for the writability-changed callback to be called again. I think with that in mind, it's simple enough to keep the code as currently is? (at least IMO here it's slightly easier to read having everything in one method body)

Reply:
Can we remove the flush first for the normal writes, in a separate PR? That would simplify this PR a bit and also ensure that not flushing is sound.

Reply:
Sure :) https://github.com/elastic/elasticsearch/pull/89647